### PR TITLE
feat: column-major streaming Parquet writer primitive

### PR DESCRIPTION
**Pull request by g-talbot**
## Summary

PR-2 of the streaming-merge stack. Lands a primitive in
`quickwit-parquet-engine/src/storage/streaming_writer.rs` that wraps
`SerializedFileWriter` directly (not `ArrowWriter`) and exposes a
per-column write API. Each column's chunk is appended and flushed
before the next column is written, so peak memory per row group is
bounded by one column's encoded chunk plus small bookkeeping (bloom
filters + page indexes) — not by the entire row group.

No production callers in this PR. PR-6 will cut the merge engine over
(primary consumer of `write_next_column_arrays`); PR-3 (last in the
stack) will cut ingest over to a single-RG writer using
`write_next_column`.

## Two write APIs

- **`write_next_column(&ArrayRef)`** — one whole arrow column per
  call. Used by ingest (PR-3), where the input is already a complete
  batch in memory.

- **`write_next_column_arrays(impl IntoIterator<Item = ArrayRef>)`** —
  next column as a stream of arrow arrays, each typically one input
  page's worth of values. Used by PR-6's merge engine, which natively
  works with arrow arrays for sort-key comparison and produces an
  array per input page as it drives the merge.

  Output page boundaries are determined by the writer's
  `data_page_size_limit` / `data_page_row_count_limit` and need not
  match the input partition; logical values round-trip. Memory is
  bounded by one input array plus the writer's internal page
  accumulator (~one output-page worth) — by **page** size, not
  column-chunk size.

### Why arrays, not raw page bytes

The arrays-input path was originally planned as a raw-page-bytes
input (`InputPage { header, bytes }`) so PR-6 could enable direct
page copy (no decompress + recompress) for columns whose pages all
come from one input. Direct page copy requires parquet 58's
`SerializedRowGroupWriter::next_column_with_factory` to be public,
which it is not (`pub(crate)`). That optimization is deferred to a
future arrow-rs upstream PR plus a follow-up here. The arrays-input
shape delivers the same per-page memory bound today using only the
public parquet API; PR-6 decodes input pages to arrow arrays
(which it would do anyway for sort-key comparison) and feeds them
directly through `write_next_column_arrays`.

## Caller contract

```text
1. start_row_group()                                       → RowGroupBuilder<'_, W>
2. RowGroupBuilder::{write_next_column | write_next_column_arrays}
   once per top-level arrow field, in arrow schema order
3. RowGroupBuilder::finish()
4. repeat (1)–(3) for additional row groups
5. StreamingParquetWriter::close()                          → ParquetMetaData
```

Out-of-order calls (too many columns, finish before all columns,
mismatched row counts) return a structured `ParquetWriteError::SchemaValidation`
rather than panicking. The compiler enforces single-RG-open via the
lifetime on `RowGroupBuilder`.

## Limitations (PR-2)

Top-level arrow fields must each map to exactly one parquet leaf —
"flat" schemas of primitive, byte-array, and dictionary types. The
metrics schema is flat in this sense. Nested types (Struct, List,
Map) are rejected at `start_row_group` with a clear error message.

The streaming writer does **not** implicitly add the `ARROW:schema`
IPC entry that `ArrowWriter` writes by default. Callers wanting that
self-describing arrow round-trip pass their `WriterProperties`
through `parquet::arrow::add_encoded_arrow_schema_to_metadata` first.

## Tests (21)

Round-trip: single RG, multi RG, nulls preserved.

Metadata identity vs `ArrowWriter` under identical `WriterProperties`:
single RG and multi RG. Asserts schema descriptor, per-RG
`sorting_columns`, all `qh.*` KV entries, per-column compression,
per-column bloom filter presence, per-column `statistics_enabled`
level, num row groups, and per-RG num_rows.

Functional contracts: chunk-level statistics propagate, bloom
filter round-trip (present + absent values), bounded memory
(pending-writers' total `memory_size` is monotone non-increasing).

Caller-contract violations: nested type rejected at start_row_group;
writing past last field; row-count mismatch across columns; finish
before all columns; empty row group still produces a readable file.

Array-stream-specific (8 new): page-stream round-trip; with nulls;
varied chunk sizes per column; single-row-per-array; empty iterator
→ zero-row column; row-count-mismatch error; too-many-columns error;
bounded memory under array streaming.

## Position in stack (execution order)

PR labels (PR-1..PR-7) are stable IDs that match opened PR
descriptions. The execution order below is updated 2026-05-05 to
put PR-3 (ingest cutover) last — the OOM concern lives in
compaction, not ingest.

| Order | PR | Status | Description |
|-------|----|--------|-------------|
| 1 | PR-1 | open ([#6377](https://github.com/quickwit-oss/quickwit/pull/6377)) | Page-level stats + `qh.rg_partition_prefix_len` marker |
| 2 | **PR-2** | **this PR** | Streaming column-major writer primitive |
| 3 | PR-4 | open ([#6386](https://github.com/quickwit-oss/quickwit/pull/6386)) | Streaming column-major reader, page-level |
| 4 | PR-5 | next | Legacy multi-RG input adapter |
| 5 | PR-6 | | Streaming merge engine; consumes PR-2 + PR-4 |
| 6 | PR-7 | | Wire `ParquetMergeExecutor`; delete downloader |
| 7 | PR-3 | last | Ingest writer cuts over to single-RG (least critical — splits already small) |

PR-2 is independent of PR-1 review — it branches off `main`.

## Test plan

- [x] `cargo +nightly fmt --all -- --check` (per files touched)
- [x] `RUSTFLAGS="-Dwarnings --cfg tokio_unstable" cargo clippy -p quickwit-parquet-engine --tests`
- [x] `cargo doc -p quickwit-parquet-engine --no-deps`
- [x] `cargo machete`
- [x] `bash quickwit/scripts/check_license_headers.sh`
- [x] `bash quickwit/scripts/check_log_format.sh`
- [x] `cargo nextest run -p quickwit-parquet-engine --all-features` — 377 tests, all pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by mattmkim — APPROVED** [review]

File: quickwit/quickwit-parquet-engine/src/storage/streaming_writer.rs [hunk 1]
```diff
@@ -0,0 +1,1375 @@
...
+                    self.next_field_idx,
+                    field.name(),
+                    total_rows,
+                    expected,
+                )));
+            }
+        }
+
+        let chunk = writer.close()?;
+        chunk.append_to_row_group(&mut self.row_group_writer)?;
```
[the remark this answers is not in the document]
Comment (g-talbot) [reply quickwit/quickwit-parquet-engine/src/storage/streaming_writer.rs]: You're right, and thanks — this is exactly the silent weakening I'd documented in the sesh-mode skill update. The previous implementation went through `ArrowColumnWriter`, whose internal `ArrowPageWriter` accumulates pages in a `SharedColumnChunk` until `close + append_to_row_group`. So even with one array per input page, memory still grew with the column chunk — defeating the page-bounded contract.

Fixed in `1949915db`:

- Switched `write_next_column_arrays` to `row_group_writer.next_column()` → `SerializedColumnWriter`, whose `SerializedPageWriter` flushes pages directly through `TrackedWrite` to the user-supplied sink. No `SharedColumnChunk` in the middle.
- Reimplemented the Arrow → typed-values dispatch at the call site (Boolean, Int8/16/32, UInt8/16, Int64, UInt32/64, Float32/64, Utf8/LargeUtf8, Binary/LargeBinary, Dictionary materialized via `take`) — covers the metrics schema's physical types.
- Renamed the old `test_array_stream_bounded_memory_per_column` to `test_array_stream_pending_writers_drained_per_column` to reflect what it actually tests (pre-allocated-writer cleanup, not in-flight column memory).
- Added `test_array_stream_pages_flush_incrementally`: uncompressed UInt64 with 32 KiB pages (each larger than the 8 KiB `BufWriter` inside `TrackedWrite`). Asserts per-column sink-write count matches one-write-per-page (SerializedColumnWriter's pattern) and rejects `io::copy`-at-end (ArrowColumnWriter's pattern). Doc comment is honest that this is indirect runtime evidence; the strong claim follows from `next_column()` being on the path.

Direct page copy (input bytes → output bytes, no decompress + recompress) for single-input columns is still deferred — `SerializedRowGroupWriter::next_column_with_factory` is `pub(crate)` in arrow-rs, blocking that fast path until an upstream PR lands.

File: quickwit/quickwit-parquet-engine/src/storage/streaming_writer.rs [hunk 2]
```diff
@@ -0,0 +1,1915 @@
...
+    }
+
+    /// Open a new row group. The returned [`RowGroupBuilder`] borrows
+    /// `self` for the lifetime of the row group.
+    ///
+    /// Returns an error if the arrow schema contains a field that maps
+    /// to more or fewer than one parquet leaf column.
+    pub(crate) fn start_row_group(&mut self) -> Result<RowGroupBuilder<'_, W>, ParquetWriteError> {
+        let column_writers = self.factory.create_column_writers(self.next_rg_idx)?;
+        if column_writers.len() != self.arrow_schema.fields().len() {
```
[the remark this answers is not in the document]
Comment (g-talbot) [reply quickwit/quickwit-parquet-engine/src/storage/streaming_writer.rs:134]: Fixed in `258012cea`.

Sketches must work end-to-end *and* the page-bounded contract must hold. The right answer is to support `List<T>` / `LargeList<T>` (with non-nullable outer + non-nullable inner over a flat numeric primitive) on the same page-bounded `SerializedColumnWriter::write_batch` path the flat primitives use, with hand-computed Dremel def/rep levels (`max_def = 1`, `max_rep = 1`):

| case | def | rep | value? |
|---|---|---|---|
| empty list | 0 | 0 | — |
| non-empty list, first element | 1 | 0 | yes |
| non-empty list, subsequent element | 1 | 1 | yes |

Memory stays bounded by one in-flight page — same contract as the flat path. Pages flush directly to the file sink as `data_page_size_limit` / `data_page_row_count_limit` are reached.

`start_row_group`'s leaf-count check is also clarified: it now rejects only **multi-leaf** nested types (e.g. `Struct` with several primitive children) up front. Single-leaf nested goes through the new list path. Other nested shapes (nullable inner, nullable outer, `Map`, `FixedSizeList`, etc.) return `SchemaValidation` with a clear message.

New regression tests:
- `test_list_uint64_round_trip_through_array_stream` — DDSketch `counts` shape: multiple input arrays, variable-length lists, edge values (`u64::MAX`, empty list), row-by-row value equality.
- `test_list_uint64_pages_flush_incrementally` — mirrors `test_array_stream_pages_flush_incrementally`: 32 768 rows × 4096-row page limit → ≥ 8 sink writes during the column write. A column-chunk-buffered fallback would not pass this assertion; that's the page-bounded property under test.
- `test_multi_leaf_nested_type_rejected_at_start_row_group` (renamed from `test_nested_type_rejected_at_start_row_group`).

25/25 streaming_writer tests + 448/448 crate-wide pass.

(My initial fix routed nested through `ArrowColumnWriter`, which is column-chunk-buffered — that silently weakened the page-bounded contract. Reverted; this comment describes the corrected approach.)
